apps/contact-service/src/routes/contacts/importContacts.ts

```diff
@@ -31,21 +45,61 @@ export const importContacts = Handler.make(
           }))
         )
 
+        yield* _(Effect.log('Contacts before', contactsBefore))
+        yield* _(Effect.log('Contacts to insert', contactsToInsert))
+
+        const newContacts = Array.differenceWith<{
+          hashFrom: HashedPhoneNumber
+          hashTo: HashedPhoneNumber
+        }>((a, b) => a.hashFrom === b.hashFrom && a.hashTo === b.hashTo)(
+          contactsToInsert,
+          contactsBefore
+        )
+        yield* _(Effect.log('New contacts', newContacts))
+
+        const contactsCountToReachQuota = yield* _(
+          importContactsQuota.getImportedContactsCountToReachQuotaForPhoneNumber(
+            security.hash
+          )()
+        )
+
+        if (
+          Option.isSome(user) &&
+          user.value.initialImportDone &&
+          newContacts.length > contactsCountToReachQuota
+        ) {
+          return yield* _(Effect.fail(new ImportContactsQuotaReachedError()))
+        }
+
+        yield* _(contactDb.deleteContactsByHashFrom(security.hash))
+
         yield* _(
           Effect.forEach(contactsToInsert, contactDb.insertContact, {
             batching: true,
           })
         )
 
-        const newContacts = Array.difference(contactsToInsert, contactsBefore)
-
         yield* _(Effect.log('New contacts', newContacts))
 
         yield* _(
           DashboardReportsService,
           Effect.flatMap((service) => service.reportContactsImported())
         )
 
```

Comment:
Put the logic for checking if quota was exceeded and trowing error into ImportContactsQuotaService